File: sorting/golang/bubble.go
package main

import (
	"fmt"
)

func bubble_sort(s []int) []int {
	count := 0
	n := len(s)
	for i := 0; i < n; i++ {
		for j := 1; j < n-i; j++ {
			count++
			if s[j] < s[j-1] {
				s[j], s[j-1] = s[j-1], s[j]
			}
		}
	}
	fmt.Println(count)
	return s
}

func bubble_sort2(s []int) []int {
	count := 0
	n := len(s)
	for i := 0; i < n; i++ {
		flag := true
		for j := 1; j < n-i; j++ {
			count++
			if s[j] < s[j-1] {
				s[j], s[j-1] = s[j-1], s[j]
				flag = false
			}
		}
		if flag {
			break
		}
	}
	fmt.Println(count)
	return s
}

func bubble_sort3(s []int) []int {
	count := 0
	n := len(s)
	last := n
	for i := 0; i < n; i++ {
		flag := true
		temp := last
		for j := 1; j < temp; j++ {
			count++
			if s[j] < s[j-1] {
				s[j], s[j-1] = s[j-1], s[j]
				flag = false
				last = j
			}
		}
		if flag {
			break
		}
	}
	fmt.Println(count)
	return s
}

func main() {
	s := []int{1, 3, 9, 0, 2, 4, 7}
	fmt.Println(s, "---default")
	// fmt.Println(bubble_sort(s), "---bubble")
	// fmt.Println(bubble_sort2(s), "---bubble2")
	fmt.Println(bubble_sort3(s), "---bubble3")
}

File: sorting/python/bubble.py
#!/usr/local/bin/python2
# -*- coding: UTF-8 -*-

def bubble_sort(arr):
    n = len(arr)
    for i in range(n - 1):
        for j in range(1, n - i):
            if arr[j - 1] > arr[j]:
                arr[j - 1], arr[j] = arr[j], arr[j - 1]
    return arr

def bubble_sort2(arr):
    n = len(arr)
    for i in range(n - 1):
        flag = True
        for j in range(1, n - i):
            if arr[j - 1] > arr[j]:
                arr[j - 1], arr[j] = arr[j], arr[j - 1]
                flag = False
        if flag:
            break
    return arr

def bubble_sort3(arr):
    n = len(arr)
    last = n
    for i in range(n - 1):
        flag = True
        for j in range(1, last):
            if arr[j - 1] > arr[j]:
                arr[j - 1], arr[j] = arr[j], arr[j - 1]
                flag = False
                last = j
        if flag:
            break
    return arr

if __name__ == "__main__":
    arr = [1, 3, 9, 0, 2, 4, 7, 8, 6, 5]

    print arr, "---default"
    # print bubble_sort(arr), "---bubble"
    # print bubble_sort2(arr), "---bubble2"
    print bubble_sort3(arr), "---bubble3"
